core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java

```diff
@@ -1613,21 +1642,24 @@ public void replaceAll(BiFunction<? super K, ? super V, ? extends V> function) {
             if (oldValue != null) {
                V newValue = function.apply(k, oldValue);
                Objects.requireNonNull(newValue, NULL_VALUES_NOT_SUPPORTED);
-               if (hasListeners) {
-                  cacheNotifier.notifyCacheEntryModified(k, newValue, defaultMetadata, oldValue, oldEntry.getMetadata(),
-                        true, ImmutableContext.INSTANCE, null);
+               if (aggregateCompletionStage != null) {
+                  aggregateCompletionStage.dependsOn(cacheNotifier.notifyCacheEntryModified(k, newValue, defaultMetadata, oldValue, oldEntry.getMetadata(),
+                        true, ImmutableContext.INSTANCE, null));
                }
                ref.set(k, newValue, oldValue, oldEntry.getMetadata());
                return factory.update(oldEntry, newValue, defaultMetadata);
             } else {
                return null;
             }
          });
-         if (hasListeners) {
-            cacheNotifier.notifyCacheEntryModified(ref.getKey(), ref.getNewValue(), defaultMetadata, ref.getOldValue(),
-                  ref.getOldMetadata(), false, ImmutableContext.INSTANCE, null);
+         if (aggregateCompletionStage != null) {
+            aggregateCompletionStage.dependsOn(cacheNotifier.notifyCacheEntryModified(ref.getKey(), ref.getNewValue(), defaultMetadata, ref.getOldValue(),
+                  ref.getOldMetadata(), false, ImmutableContext.INSTANCE, null));
```

Comment:
Hrmm, good point.

Unfortunately, with the way SimpleCache works right now, there is no way to have the pre notification be non blocking and have it prevent the write. Unfortunately relying on a sync lock callback like compute doesn't really jive with that.

We may have to revisit this later then. I think if we want to keep this up we should probably investigate that in https://issues.jboss.org/browse/ISPN-3643

Reply:
ISPN-3643 is weird, I was sure we already behave that way (and `ListenerExceptionTest` seems to confirm that). I would be fine with `SimpleCache` blocking to wait for the the pre listeners' CompletionStages, because we're never going to change it to acquire locks without blocking, either.